products/endpoints/backend/api.py

```diff
@@ -473,20 +539,57 @@ def _execute_inline_endpoint(self, endpoint: Endpoint, data: EndpointRunRequest,
 
     @extend_schema(
         request=EndpointRunRequest,
-        description="Execute endpoint with optional materialization.",
+        description="Execute endpoint with optional materialization. Supports version parameter, runs latest version if not set.",
     )
     @action(methods=["GET", "POST"], detail=True)
     def run(self, request: Request, name=None, *args, **kwargs) -> Response:
-        """Execute endpoint with optional parameters."""
+        """Execute endpoint with optional parameters.
+
+        Query Parameters:
+            version (int, optional): Specific version to execute. Defaults to latest.
+        """
         endpoint = get_object_or_404(Endpoint, team=self.team, name=name, is_active=True)
         data = self.get_model(request.data, EndpointRunRequest)
 
-        use_materialized = self._should_use_materialized_table(endpoint, data)
+        version_param = request.query_params.get("version") or request.data.get("version")
+        version_number = None
+
+        if version_param is not None:
+            try:
+                version_number = int(version_param)
+            except (ValueError, TypeError):
+                return Response(
+                    {"error": f"Invalid version parameter: {version_param}"}, status=status.HTTP_400_BAD_REQUEST
+                )
+
+        version_obj = None
+        try:
+            version_obj = endpoint.get_version(version_number)
+        except EndpointVersion.DoesNotExist:
+            if version_number is not None:
+                return Response(
+                    {
+                        "error": f"Version {version_number} not found for endpoint '{name}'",
+                        "current_version": endpoint.current_version,
+                    },
+                    status=status.HTTP_404_NOT_FOUND,
+                )
+
+        # Only the latest version is materialized
+        use_materialized = version_number is None and self._should_use_materialized_table(endpoint, data)
 
         if use_materialized:
-            return self._execute_materialized_endpoint(endpoint, data, request)
+            result = self._execute_materialized_endpoint(endpoint, data, request)
         else:
-            return self._execute_inline_endpoint(endpoint, data, request)
+            # Use version's query if available, otherwise use endpoint.query
+            query_to_use = version_obj.query if version_obj else None
```

Comment:
if you write `query_to_use = version_obj.query if version_obj else endpoint.query.copy()` here, you may skip the endpoint from _execute_inline_endpoint - it becomes simpler